queue/dependencytracker.js

```diff
@@ -0,0 +1,279 @@
+
+    // Create TaskRequirement entries, each entry implies that taskId is blocked
+    // by requiredTaskId. This relation is used to track if a taskId is blocked.
+    await Promise.all(task.dependencies.map(requiredTaskId => {
+      return this.TaskRequirement.create({
+        taskId:           task.taskId,
+        requiredTaskId,
+        expires:          task.expires,
+      }, true);
+    }));
+
+    // Create TaskDependency entries, each entry implies that taskId is required
+    // by dependentTaskId. This relation is used so taskId can find dependent
+    // tasks when it is resolved.
+    await Promise.all(task.dependencies.map(requiredTaskId => {
+      return this.TaskDependency.create({
+        taskId:           requiredTaskId,
+        dependentTaskId:  task.taskId,
+        expires:          task.expires,
+        relation:         task.dependencyRelation,
+      }, true);
+    }));
+
+    // Load all task dependencies to see if they have been resolved.
+    // We will also check for missing and expiring dependencies.
+    let missing = [];         // Dependencies that doesn't exist
+    let expiring = [];        // Dependencies that expire before deadline
+    let anySatisfied = false; // Track if any dependencies were satisfied
+    await Promise.all(task.dependencies.map(async (requiredTaskId) => {
+      let requiredTask = await this.Task.load({taskId: requiredTaskId}, true);
+
+      // If task is missing, we should report and error
+      if (!requiredTask) {
+        return missing.push(requiredTaskId);
+      }
+
+      // Check if requiredTask expires before the deadline
+      if (task.deadline.getTime() > requiredTask.expires.getTime()) {
+        return expiring.push(requiredTaskId);
+      }
+
+      // Check if requiredTask is satisfied
+      let state = requiredTask.state();
+      if (state === 'completed' || (task.dependencyRelation === 'on-resolved' &&
+          (state === 'exception' || state === 'failed'))) {
+        // If a dependency is satisfied we delete the TaskRequirement entry
+        await this.TaskRequirement.remove({
+          taskId:         task.taskId,
+          requiredTaskId,
+        }, true);
+        // Track that we've deleted something, now we must check if any are left
+        // afterward (using isBlocked)
+        anySatisfied = true;
+      }
+    }));
+
+    // If we found some missing dependencies we're done, createTask should
+    // clearly return an error
+    if (missing.length > 0 || expiring.length > 0) {
+      // Construct explanatory error message
+      let msg = '';
+      if (missing.length > 0) {
+        msg += '`task.dependencies` references non-existing tasks: \n';
+        msg += missing.map(taskId => {
+          return ' * ' + taskId + ','
+        }).join('\n') + '\n';
+        msg += 'All taskIds in `task.dependencies` **must** exist\n';
+        msg += 'before the task is created.\n';
+      }
+      if (expiring.length > 0) {
```

Comment:
We make this assertion so that we're free to schedule a task all the way up to its deadline, right? Without this assertion we'll sometimes still be able to succeed, but there would be situations where things don't work anymore, right?